src/merge_operator.rs

```diff
@@ -0,0 +1,255 @@
+}
+
+pub(crate) type MergeOperatorType = Arc<dyn MergeOperator + Send + Sync>;
+
+pub(crate) struct MergeOperatorIterator<T: KeyValueIterator> {
+    merge_operator: MergeOperatorType,
+    delegate: T,
+    /// Entry from the delegate that we've peeked ahead and buffered.
+    buffered_entry: Option<RowEntry>,
+    /// Whether to merge entries with different expire timestamps.
+    merge_different_expire_ts: bool,
+}
+
+#[allow(unused)]
+impl<T: KeyValueIterator> MergeOperatorIterator<T> {
+
+    pub(crate) fn new(merge_operator: MergeOperatorType, delegate: T) -> Self {
+        Self {
+            merge_operator,
+            delegate,
+            buffered_entry: None,
+            merge_different_expire_ts: true,
+        }
+    }
+}
+
+impl<T: KeyValueIterator> KeyValueIterator for MergeOperatorIterator<T> {
+    async fn next_entry(&mut self) -> Result<Option<RowEntry>, SlateDBError> {
+        let next_entry = match self.buffered_entry.take() {
+            Some(entry) => Some(entry),
+            None => self.delegate.next_entry().await?,
+        };
+        if let Some(entry) = next_entry {
+            match &entry.value {
+                // A mergeable entry, we need to accumulate all mergeable entries
+                // ahead for the same key and merge them into a single value.
+                ValueDeletable::Merge(merge_value) => {
+                    let current_key = entry.key.clone();
+                    let mut current_value = merge_value.clone();
+                    let mut max_create_ts = entry.create_ts;
+                    let mut min_expire_ts = entry.expire_ts;
+                    let mut max_seq = entry.seq;
+
+                    // Keep looking ahead and merging as long as we find mergeable entries
+                    loop {
+                        let next = self.delegate.next_entry().await?;
+                        match next {
+                            Some(next_entry)
+                                if next_entry.key == current_key
+                                    && (self.merge_different_expire_ts
+                                        || entry.expire_ts == next_entry.expire_ts) =>
+                            {
+                                // Update timestamps
+                                max_create_ts =
+                                    merge_options(max_create_ts, next_entry.create_ts, i64::max);
+                                min_expire_ts =
+                                    merge_options(min_expire_ts, next_entry.expire_ts, i64::min);
+                                // Update sequence number
+                                max_seq = max_seq.max(next_entry.seq);
+
+                                match next_entry.value {
+                                    ValueDeletable::Value(value) => {
+                                        // Final merge with a regular value
+                                        let merged =
```

Comment:
I'm confused by this one. If we have:

```
t0, k1, merge1
t1, k1, merge2
t2, k1, value
```
Shouldn't the iterator return `value` since it overwrites all previous `k1`'s?

The way I read the code, it would merge `merge1`, `merge2`, and `value` and return that merged result.

Reply:
I'm working on more complete and expressive tests right now, but in general we need to make sure all KeyValueIterators hold up to a following order:
- Primary Sorted by Key
- Secondary Sorted by seq_num in reverse order (latest entry for the key always come first); that's also how data should be laid out in SST file

Related discussion:

https://github.com/slatedb/slatedb/pull/389#discussion_r1878011239

I've asked ChatGPT if it's aligned with rocksdb approach:

```
2. Sorting in SST Files
	•	SST files are sorted by internal keys, not just the user key.
	•	Since the sequence number is appended as part of the internal key (in reverse order), it ensures that:
	•	Newer versions of a key (with higher sequence numbers) appear before older versions of the same key within the sorted order of the SST file.
	•	This is achieved because the sequence number is stored in a descending order as part of the key’s internal representation.
```

---

So back to the example above, merges are the latest entries and we keep merging until we hit a terminal entry (value / tombstone).

Any thoughts?

Reply:
> So back to the example above, merges are the latest entries and we keep merging until we hit a terminal entry (value / tombstone).

+1

merge should take the seqnum into consideration. if we got two entries with two seq numbers, we should only keep the latest one 😲

Reply:
Ahh! I forgot that values for the same key are sorted in reverse order in an SST. Ok, this makes sense.